mother/src/drive.c

```diff
@@ -0,0 +1,252 @@
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <app_version.h>
+#include <math.h>
+
+struct DiffDriveOdometryConfig {
+	int timestamp;
+	float x, y, heading, wheel_separation, left_wheel_radius, right_wheel_radius;
+	int velocity_rolling_window_size;
+};
+struct DiffDriveOdometry {
+	const struct DiffDriveOdometryConfig config;
+	int timestamp;
+	float x, y, heading, left_wheel_old_pos, right_wheel_old_pos, linear, angular;
+	struct FloatRollingMeanAccumulator *linear_accumulator, *angular_accumulator;
+};
+
+int diffdrive_odometry_update_from_velocity(struct DiffDriveOdometry *odom, float left_vel,
+					    float right_vel, const int time);
+
+struct DiffDriveTwist {
+	float linear_x;
+	float angular_z;
+};
+
+enum DiffDriveUpdateType {
+	POSITION_FEEDBACK,
+	VELOCITY_FEEDBACK
+};
+
+struct DiffDriveConfig {
+	float wheel_separation;
+	int wheels_per_side; // actually 2, but both are controlled by 1 signal
+	float wheel_radius;
+
+	float wheel_separation_multiplier;
+	float left_wheel_radius_multiplier;
+	float right_wheel_radius_multiplier;
+	int update_type;
+};
+
+struct DiffDrive {
+	const struct DiffDriveConfig config;
+	int previous_update_timestamp;
+	const struct DiffDriveOdometryConfig odometry_config;
+	struct DiffDriveOdometry *odometry;
+	int (*feedback_callback)(float *feedback_buffer, int buffer_len, int wheels_per_side);
+	int (*velocity_callback)(const float *velocity_buffer, int buffer_len, int wheels_per_side);
+};
+
+void *diffdrive_init(struct DiffDriveConfig *config)
+{
+	// TODO: Make a constructor here
+	return (void *)0xDEADBEEF;
+}
+
+void diffdrive_update(struct DiffDrive *drive, struct DiffDriveTwist command)
+{
+	// Get time since last update
```

Comment:
We have two apis:
- `k_uptime_get()` - simply returns uptime since boot
- `k_uptime_delta(ref)` - returns the delta value from ref and updates ref to current uptime

you mention time since last update, for that i need time when last update occured. If you just need uptime, you can use `k_uptime_get`

Reply:
Ended up using `k_uptime_delta` in
https://github.com/waseemR02/kyvernitis/blob/f26e5e8019877029e4b784c2bbb58382264ed5b4/mother/src/drive.c#L69-L72